### ISSUE TITLE
[Java] Import static * is flagged as an error if even one method or constant is unused in the using class. 

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#1870]
**Affects PMD Version: 6.15 ** 

**Rule: UnusedImports**

**Description:**
com/howtodoinjava/hashing/password/demo/bcrypt/BCrypt.java:17:   Avoid unused imports such as 'com.howtodoinjava.hashing.password.demo.bcrypt.BCryptConstants'

I receive the above message while running PMD on my project Facade2 on my Github repository https://github.com/Fernal73/LearnJava

BCryptConstants is a refactored class to reduce the size of the original class Bcrypt since PMD flagged it as too large. The constants used are relocated to this CONSTANTS class. 

Newer coding standards ddictate that static imports are the way to use Constants in java as opposed to placing them in an interface class. 

The imports have to be static. 
However, PMD flags static imports as unused if even one method or constant is unused. 
If I were to import every constant, it would make for a very long list of import statements. That's hardly warranted, in my opinion. Using that many static imports might clash with the other PMD rule, TooManyStaticImports. 

The rule ought to be relaxed by allowing for a configurable parameter that specifies the maximum number of unused imports allowed for static * imports. 

Note that this is very specific to static * imports only. For classes in packages, the rule is perfectly fine. 

**Running PMD through:** *[CLI | Ant] 
USE the script pmd (from root directory) with Project directory as command line parameter or run ant pmd under the project directory. You will have to run the root directory's build.xml first to generate the classpath before running either of the above. Ivy downloads the jars for PMD and generates a classpath file used by the script and every project's pmd task. 
*

**Comment by adangel** [comment]
Thanks for creating the issue.

> However, PMD flags static imports as unused if even one method or constant is unused.

That should not be the case. If it is that way, then this is a false positive. PMD should only flag the import as unused, if it is totally unused. One usage should be enough to count as usage.

Looking at [BCryptConstants](https://github.com/Fernal73/LearnJava/blob/master/Facade2/com/howtodoinjava/hashing/password/demo/bcrypt/BCryptConstants.java) I see, that all constants are package-private. Maybe we don't consider this case in PMD and only check for usages of public fields?

Could you try this theory and just make one constant public and check, whether the unused import violation is still reported?

> Newer coding standards dictate that static imports are the way to use Constants in java as opposed 
> to placing them in an interface class.

Just to clarify, why (mis-)using interfaces for constants is not good design: This leads to implementing this "constants interface" just to have access to the constants. The constants are usually a implementation detail, but exposing this as a interface makes this implementation detail leak into the API of the class. Also, if a class implements a interface, it usually means, that the class is of this type. But what type would a interface be, that only contains constants (and no methods)? You could cast the class to the interface but it is of no use at all.
This is described in more detail in Item 19 (Use interfaces only to define types) in Effective Java. An excerpt is here: http://thefinestartist.com/effective-java/19

Using static imports is just syntactic sugar to mimic the old simple usage of constants without needing to write the class, where the constants are defined.

> The rule ought to be relaxed by allowing for a configurable parameter that specifies the maximum 
> number of unused imports allowed for static * imports.

This should not be necessary as a single usage should be enough. You don't need to use all constants when using wildcard imports.

**Comment by Fernal73** [comment]
Thanks, Andreas.

I made one field public and tried the star import. But it still flags the
error.

I have no problems with static constant imports. In fact, that works for
me. Though, I did cavil initially.

If it's a problem with the fields being package private, PMD ought to look
into that ASAP.

Frankly, if constants are a public contract if defined in an interface, so
are public constants since they're accessible publicly. I'd prefer to have
such fields package-private if they can't be part of the class itself.

**Comment by Fernal73** [comment]
Also, I'm using all the constants in the class since they've simply been
moved into BCryptConstants.java. There simply shouldn't be an UnusedImports
violation. False +ve, as you said.

**Comment by adangel** [comment]
The problem is a incorrect auxclasspath. The correct PMD task definition in your ant-file should look like this:

```xml
  <target name="pmd">
    <pmd cacheLocation="${pmd.cacheFile}" minimumPriority="${pmd.minimumPriority}" suppressMarker="${pmd.suppressMarker}">
      <sourceLanguage name="java" version="${ant.build.javac.source}"/>
      <ruleset>../${pmd.config}</ruleset>
      <auxclasspath>
        <pathelement path="${basedir}" />
      </auxclasspath>
      <formatter type="${pmd.reportFormat}" toFile="${pmd.output}" toConsole="${pmd.toConsole}">
      </formatter>
      <fileset dir="${basedir}">
        <include name="**/*.java"/>
      </fileset>
    </pmd>
  </target>
```

Note: the auxclasspath is similar like any other java classpath. This means, you specify directories and/or jar files, where compiled classes are located. You do not need to specify single class files.

Also, you seem to have the auxclasspath twice in the target definition. Only one is sufficient.

With the above target, I got only one rule violation: "BCrypt.java:350:	The method 'retrieveOffsetMinor(String)' has a cyclomatic complexity of 10."

**Comment by adangel** [comment]
btw. package-private issues are #925 and #1209 and are already fixed.

**Comment by Fernal73** [comment]
This is the second time this rule failed (for me) with the auxiliary
classpath the issue. Do you have a list of rules that will need the
project's generated class files to be present in the classpath for the
rules to work as expected?

It would also help to have the auxclasspath element documented with an
example in the PMD ant task documentation.

Thank you very much for your time.

Regards,
Linus.

**Comment by Fernal73** [comment]
I'm assuming the fileset specification didn't work because only the short
class names were looked at, not the full package directory structure. Is
that what's expected?

**Comment by Fernal73** [comment]
OK. I see it now. The classpath would have every class file's path.
And a ClassLoader won't pick them up?
Hmm...

**Comment by adangel** [comment]
Exactly - it behaves like a normal java [classpath](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/classpath.html). Except for the wildcard classpath feature - I'm not sure, whether we correctly support this for auxclasspath.

> Do you have a list of rules that will need the project's generated class files to be present in the 
> classpath for the rules to work as expected?

Not all rules require it, but more and more do. Rules are regularly improved to use type resolution since that helps to avoid false positives.
So it is simpler to always execute PMD with the correct auxclasspath. When executing PMD using maven and the maven-pmd-plugin, this is done automatically. The same with gradle. With ant, you are on your own, because there is no dependency management integrated into the build tool.

**Comment by adangel** [comment]
I'm closing this issue, since the problem was configuration/missing auxclasspath. I've improved the documentation about the ant task, see #1871.

**Comment by Fernal73** [comment]
Great. Thanks.

Just found out that you did have some kind of list earlier:

https://stackoverflow.com/q/43800383/3924108

Regards,
Linus.
